File: mayonez-base/src/main/java/mayonez/renderer/shader/ShaderType.java
package mayonez.renderer.shader;

import mayonez.application.*;

import static org.lwjgl.opengl.GL20.GL_FRAGMENT_SHADER;
import static org.lwjgl.opengl.GL20.GL_VERTEX_SHADER;

/**
 * The type of .glsl shader stage used by the GPU.
 *
 * @author SlavSquatSuperstar
 */
@UsesBackend(Backend.GL)
enum ShaderType {

    /**
     * A vertex shader that transforms an image to its screen position.
     */
    VERTEX(GL_VERTEX_SHADER),

    /**
     * A fragment (pixel) shader that calculates the color and texture of each
     * individual pixel of an image.
     */
    FRAGMENT(GL_FRAGMENT_SHADER);

    /**
     * The shader ID in OpenGL.
     */
    final int glShaderType;
    final String name;

    ShaderType(int glShaderType) {
        this.glShaderType = glShaderType;
        this.name = name().toLowerCase();
    }

    @Override
    public String toString() {
        return name;
    }

}

File: mayonez-base/src/main/kotlin/mayonez/renderer/shader/ShaderType.kt
package mayonez.renderer.shader

import mayonez.graphics.*
import mayonez.util.*
import org.lwjgl.opengl.GL20

/**
 * The type of shader used by the GPU.
 *
 * @author SlavSquatSuperstar
 */
@UsesEngine(EngineType.GL)
internal enum class ShaderType(
    /** The shader ID in OpenGL. */
    val glShaderType: Int
) {

    /** A vertex shader that transforms an image to its screen position. */
    VERTEX(GL20.GL_VERTEX_SHADER),

    /**
     * A fragment (pixel) shader that calculates the color and texture of each
     * individual pixel of an image.
     */
    FRAGMENT(GL20.GL_FRAGMENT_SHADER);

    companion object {
        /**
         * Returns the [ShaderType] constant matching the given name,
         * case-insensitive.
         *
         * @param shaderName the name of the shader program
         * @return the shader type
         * @throws IllegalArgumentException if the shader name is invalid
         */
        @JvmStatic
        @Throws(IllegalArgumentException::class)
        fun findWithName(shaderName: String?): ShaderType {
            return StringUtils.findWithName(entries, shaderName)
                ?: throw IllegalArgumentException("Unexpected shader type \"$shaderName\"")
        }

    }

}